### DISCUSSION TITLE
How is metadata handled with already scrobbled tracks + multiple APIs authenticated?

### DISCUSSION THREAD

**Discussion by gabeklavans** [krateng/maloja#386] [General]
So I had an artist and some albums with the wrong images. I figure it's just coming from Musicbrainz or something and there wasn't enough info on the artist, no big deal. I know that Last.fm and Spotify had more accurate images, so I threw some API keys in the settings page of my Maloja instance, restarted the container, but the images stayed the same. I see in the container that all my API keys were picked up:
```
maloja-maloja-1  | Various external services can be used to display images. If not enough of them are set up, only local images will be used.
maloja-maloja-1  |      Last.fm API Key found.
maloja-maloja-1  |      Spotify API ID found.
maloja-maloja-1  |      Spotify API Secret found.
maloja-maloja-1  |      TheAudioDB API Key found.
```
This raised a couple questions for me:
- how are images handled for already-scrobbled tracks? Is the image downloaded on first scrobble and then can't be changed unless manually updated in admin mode?
  - I see there's a setting to change how metadata is decided in the settings (first, last, majority), does this mean future scrobbles will potentially use the correct metadata now that I have my keys in?
- when I have multiple APIs authenticated like I do, how does Maloja decide which one to get metadata from?
- if my hypotheses in the first bullet are true, would it be possible to add a "refresh metadata" button to the UI?

**Comment by gabeklavans** [comment]
Okay, I'm not sure what triggered it, but I saw in the logs at some point that it started trying to get metadata for all my scrobbles from all the authenticated APIs I had listed, in addition to Musicbrainz and Deezer. I couldn't detect an exact logic for when it would give up on trying to get metadata for certain songs, but that almost answers my first questions. The order in which Maloja tries the APIs is probably just hardcoded, and I'm not sure how it decides which one to go with (I'd assume the first hit).

**Comment by duckcantcode** [comment]
I had the same question. Can you join [the Unofficial Maloja Discord server](https://discord.gg/rbVUC5Rd) so we can discuss it?
